lib/AST/ASTMangler.cpp

```diff
@@ -297,33 +301,41 @@ std::string ASTMangler::mangleKeyPathSetterThunkHelper(
   if (isa<SubscriptDecl>(property)) {
     // Subscripts can be generic, and different key paths could capture the same
     // subscript at different generic arguments.
-    for (auto &sub : subs) {
-      appendType(sub);
+    for (auto sub : subs.getReplacementTypes()) {
+      appendType(sub->mapTypeOutOfContext()->getCanonicalType());
     }
   }
   appendOperator("Tk");
+  if (expansion == ResilienceExpansion::Minimal)
+    appendOperator("q");
   return finalize();
 }
 
 std::string ASTMangler::mangleKeyPathEqualsHelper(ArrayRef<CanType> indices,
-                                                  GenericSignature *signature) {
+                                                  GenericSignature *signature,
+                                                  ResilienceExpansion expansion) {
   beginMangling();
   for (auto &index : indices)
     appendType(index);
   if (signature)
     appendGenericSignature(signature);
   appendOperator("TH");
+  if (expansion == ResilienceExpansion::Minimal)
+    appendOperator("q");
```

Comment:
Why do we need a different symbol based on whether this is for serialization?  Isn't this a problem for literally every inlinable function?

Reply:
Imagine we reference the same keypath twice, once in inlinable context and once in an ordinary function. Since these thunks have shared linkage the first one must be serialized. We could just say that all keypath thunks are serialized but then we’d have to use a slightly less efficient access pattern sometimes. Maybe this is overkill but none of this is ABI anyway.

Reply:
For comparison, reabstraction and curry thunks are always serializable; generic specializations are serializable if the caller is, so there we also mangle it into the name.

Reply:
Hmm, alright.